Add an intentionally failing test

vihu opened the pull request:
Summary
----
Hi @tom-whitehead, I have added an intentionally failing test for clustering when using lat/lon with haversine distances.
I'm trying to figure it out but could use an extra set of eyes if you wouldn't mind.

My expectation here is that there should be one cluster with three sets of points (label:  `0`) and one point which should be considered noise (label: `-1`) but instead we get all four points classified as noise, each with label = -1. I'd appreciate if you could give me some pointers on how to debug this further. Thanks!

vihu committed:
Add an intentionally failing test

tom-whitehead commented:
Hey @vihu, yeah something looks a little off. I will need to do a bit more debugging when I get a chance to figure out what is going on.

tom-whitehead commented:
@vihu I think it's possible that I introduced a bug recently when I did some refactoring. I need to spend a bit more time working on it.

vihu committed:
Add some more debugging

vihu commented:
> @vihu I think it's possible that I introduced a bug recently when I did some refactoring. I need to spend a bit more time working on it.

Oh? Well, I was doing some debugging as well:

```
running 1 test
+++M1 validated input data
+++M3 dist_metric: Haversine
+++M4 dist_matrix: [[0.0, 170.91660568199816, 2069.663951645396, 6729.372389296802], [170.91660568199816, 0.0, 1900.842479345853, 6688.069678351526], [2069.663951645396, 1900.842479345853, 0.0, 6827.42431860942], [6729.372389296802, 6688.069678351526, 6827.42431860942, 0.0]]
+++M5 core_distances: [0.0, 0.0, 0.0, 0.0]
+++M2 core_distances: [0.0, 0.0, 0.0, 0.0]
result: [-1, -1, -1, -1]
noise_count: 4
test tests::test_failing_haversine_cluster ... FAILED

failures:

failures:
    tests::test_failing_haversine_cluster

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 8 filtered out; finished in 0.00s


--- STDERR:              hdbscan tests::test_failing_haversine_cluster ---
thread 'tests::test_failing_haversine_cluster' panicked at src/lib.rs:1115:17:
assertion `left == right` failed: Should have 1 noise point
  left: 4
 right: 1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

   Canceling due to test failure
------------
     Summary [   0.003s] 1 test run: 0 passed, 1 failed, 8 skipped
        FAIL [   0.002s] hdbscan tests::test_failing_haversine_cluster
error: test run failed
```

The `core_distances` vec seems obviously incorrect there right?

tom-whitehead commented:
@vihu No I think that's expected when `min_samples` is 1.  For a neighbours algorithm, the 1st neighbour is the point itself, so the distance is zero. 
The implication of this for HDBSCAN is that what's called the mutual reachability distance between any two points will just be the distance between them per the distance metric. If `min_samples` > 1, then the mutual reachability distance for two given points (a and b) will be the greater of the distance from a and b, a's core distance and b's core distance. Setting a higher `min_samples` means that points in sparser places become more isolated/distant, whereas very dense areas aren't impacted as much.

vihu committed:
Even more debugging, min_samples=2

vihu commented:
> @vihu No I think that's expected when `min_samples` is 1. For a neighbours algorithm, the 1st neighbour is the point itself, so the distance is zero. The implication of this for HDBSCAN is that what's called the mutual reachability distance between any two points will just be the distance between them per the distance metric. If `min_samples` > 1, then the mutual reachability distance for two given points (a and b) will be the greater of the distance from a and b, a's core distance and b's core distance. Setting a higher `min_samples` means that points in sparser places become more isolated/distant, whereas very dense areas aren't impacted as much.

Right, I just noticed that, thanks for the explanation, here's some more debug output:

```
running 1 test
+++M1 validated input data
+++M3 dist_metric: Haversine
+++M4 k: 2
+++M5 data: 2
+++M6 dist_matrix: [[0.0, 170.91660568199816, 2069.663951645396, 6729.372389296802], [170.91660568199816, 0.0, 1900.842479345853, 6688.069678351526], [2069.663951645396, 1900.842479345853, 0.0, 6827.42431860942], [6729.372389296802, 6688.069678351526, 6827.42431860942, 0.0]]
+++M7 core_distances: [170.91660568199816, 170.91660568199816, 1900.842479345853, 6688.069678351526]
+++M2 core_distances: [170.91660568199816, 170.91660568199816, 1900.842479345853, 6688.069678351526]
+++M8 condensed_tree: [
    CondensedNode {
        node_id: 3,
        parent_node_id: 4,
        lambda_birth: 0.00014951997333952414,
        size: 1,
    },
    CondensedNode {
        node_id: 2,
        parent_node_id: 4,
        lambda_birth: 0.0005260825191281159,
        size: 1,
    },
    CondensedNode {
        node_id: 0,
        parent_node_id: 4,
        lambda_birth: 0.005850806573239392,
        size: 1,
    },
    CondensedNode {
        node_id: 1,
        parent_node_id: 4,
        lambda_birth: 0.005850806573239392,
        size: 1,
    },
]
+++M12 cluster_id: 5, n_samples: 4
+++M11 lambda_birth: 0.0
+++M10 stabilities: {5: 0.0}
+++M9 winning_clusters: []
result: [-1, -1, -1, -1]
noise_count: 4
test tests::test_failing_haversine_cluster ... FAILED

failures:

failures:
    tests::test_failing_haversine_cluster

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 8 filtered out; finished in 0.00s


--- STDERR:              hdbscan tests::test_failing_haversine_cluster ---
thread 'tests::test_failing_haversine_cluster' panicked at src/lib.rs:1125:17:
assertion `left == right` failed: Should have 1 noise point
  left: 4
 right: 1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

   Canceling due to test failure
------------
     Summary [   0.003s] 1 test run: 0 passed, 1 failed, 8 skipped
        FAIL [   0.002s] hdbscan tests::test_failing_haversine_cluster
error: test run failed
```

I don't quite understand `cluster_id: 5, n_samples: 4`, this presumably results in the extracted lambda_birth to be 0.0 I suspect which then leads to empty winning_clusters?

tom-whitehead commented:
@vihu I have a decent idea where the issue is. When I fix it though something else breaks, so I just need to work through it. I'll try and look at it after work tonight.

tom-whitehead commented:
Hey @vihu I think I fixed a couple of issues that relate to cases when there is a single cluster. I've raised a [PR](https://github.com/tom-whitehead/hdbscan/pull/29) but I want to test it more thoroughly before I merge. 

I'm still not sure if it will fix your particular test as a) I haven't tried it yet and b) when I ran that test on a different HDBSCAN implementation it also didn't produce the expected result.

vihu commented:
> Hey @vihu I think I fixed a couple of issues that relate to cases when there is a single cluster. I've raised a [PR](https://github.com/tom-whitehead/hdbscan/pull/29) but I want to test it more thoroughly before I merge.
> 
> I'm still not sure if it will fix your particular test as a) I haven't tried it yet and b) when I ran that test on a different HDBSCAN implementation it also didn't produce the expected result.

So it turns out that your fixes in #29 does address the issues I was facing in an internal project but this particular test I added is still busted. Regardless, thanks for the fix. Maybe I have a bad test here?
